### PR TITLE
WIP Speed up deployment groups

### PR DESCRIPTION
**Pull request by gimaker**
Instead of doing (mostly no-op) work for all deployment groups in each
iteration of the rolling-update service we now only process deployment
groups that have outstanding tasks.

This commit changes behaviour in a couple of ways, including but not
limited to:
 * Hosts added during a rolling-update will be picked up and deployed
   to "at once" (as opposed to only after the rolling-update has
   finished, which was the previous behaviour).
 * A job that is undeployed or stopped manually by the user when the
   deployment group is in the "DONE" state will be not be redeployed
   or restarted until hosts are added or another rolling-update is
   triggered by the user (e.g. via the CLI).

### CODE REVIEW

**Comment by gimaker** [comment]
@rohansingh @davidxia It's a monster.

File: helios-services/src/main/java/com/spotify/helios/rollingupdate/DeploymentGroupEventFactory.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  public enum RollingUpdateReason {
+    MANUAL,
+    HOSTS_CHANGED
+  }
+
+  public Map<String, Object> rollingUpdateTaskOk(final DeploymentGroup deploymentGroup,
+                                                 final RolloutTask task) {
+    final Map<String, Object> ev = Maps.newHashMap();
+    ev.put("timestamp", System.currentTimeMillis());
```
Comment (gimaker) [inline helios-services/src/main/java/com/spotify/helios/rollingupdate/DeploymentGroupEventFactory.java]: This needs to be cleaned up

File: helios-services/src/main/java/com/spotify/helios/rollingupdate/RollingUpdateOpFactory.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.spotify.helios.common.descriptors.DeploymentGroupStatus.State.DONE;
+import static com.spotify.helios.common.descriptors.DeploymentGroupStatus.State.FAILED;
+import static com.spotify.helios.servicescommon.coordination.ZooKeeperOperations.delete;
+import static com.spotify.helios.servicescommon.coordination.ZooKeeperOperations.set;
+
+public class RollingUpdateOpFactory {
```
Comment (gimaker) [inline helios-services/src/main/java/com/spotify/helios/rollingupdate/RollingUpdateOpFactory.java]: The name probably makes no sense -- but at least this logic is easier to unit test now

File: helios-system-tests/src/main/java/com/spotify/helios/system/DeploymentGroupTest.java [hunk 3]
```diff
@@ -409,4 +420,59 @@ public void testStoppedJob() throws Exception {
     awaitTaskState(secondJobId, host, TaskStatus.State.RUNNING);
     awaitDeploymentGroupStatus(defaultClient(), TEST_GROUP, DeploymentGroupStatus.State.DONE);
   }
+
+  @Ignore
+  @Test
+  public void testRollingUpdatePerformance() throws Exception {
```
Comment (gimaker) [inline helios-system-tests/src/main/java/com/spotify/helios/system/DeploymentGroupTest.java]: This hacky test obviously needs to be fixed or deleted

File: helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java [hunk 4]
```diff
@@ -130,6 +134,10 @@ public int compare(TaskStatusEvent arg0, TaskStatusEvent arg1) {
       STRING_LIST_TYPE =
       new TypeReference<List<String>>() {};
 
+  private static final String KAFKA_TOPIC = "HeliosDeploymentGroupEvents";
```
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java]: Use a more specific variable name?

File: helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java [hunk 5]
```diff
@@ -462,12 +470,38 @@ public void removeDeploymentGroup(final String name) throws DeploymentGroupDoesN
...
 
   @Override
-  public void updateDeploymentGroupHosts(String name, List<String> hosts)
+  public void updateDeploymentGroupHosts(final String name, final List<String> hosts)
       throws DeploymentGroupDoesNotExistException {
     log.debug("updating deployment-group hosts: name={}", name);
     final ZooKeeperClient client = provider.get("updateDeploymentGroupHosts");
     try {
-      client.setData(Paths.statusDeploymentGroupHosts(name), Json.asBytes(hosts));
+      Optional<Integer> version = Optional.absent();
```
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java]: `final`s here?

File: helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java [hunk 6]
```diff
@@ -516,190 +553,160 @@ public void rollingUpdate(final DeploymentGroup deploymentGroup,
     }
   }
 
-  @Override
-  public void rollingUpdateStep(final DeploymentGroup deploymentGroup,
-                                final RolloutPlanner rolloutPlanner)
+  private List<ZooKeeperOperation> getInitRollingUpdateOps(final ZooKeeperClient client,
```
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java]: `client` is unused

File: helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java [hunk 7]
```diff
@@ -516,190 +553,160 @@ public void rollingUpdate(final DeploymentGroup deploymentGroup,
...
+          tasks, DEPLOYMENT_GROUP_EVENT_FACTORY);
+      final RolloutTask task = tasks.getRolloutTasks().get(tasks.getTaskIndex());
+      RollingUpdateOp op = processRollingUpdateTask(
+          client, opFactory, task, tasks.getDeploymentGroup());
 
-    return opsEvents;
-  }
+      if (!op.operations().isEmpty()) {
+        final List<ZooKeeperOperation> ops = Lists.newArrayList();
+        ops.add(check(Paths.statusDeploymentGroupTasks(deploymentGroupName),
```
Comment (davidxia) [inline helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java]: How does this version thing work?

File: helios-services/src/test/java/com/spotify/helios/rollingupdate/RollingUpdateOpFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+public class RollingUpdateOpFactoryTest {
+
+  private static final DeploymentGroup DEPLOYMENT_GROUP = DeploymentGroup.newBuilder()
+      .setName("my_group")
+      .setRolloutOptions(RolloutOptions.newBuilder().build())
+      .build();
+  private static final Map<String, Object> DONE_EVENT =
```
Comment (davidxia) [inline helios-services/src/test/java/com/spotify/helios/rollingupdate/RollingUpdateOpFactoryTest.java]: Unused vars

**Comment by davidxia** [comment]
Replaced by https://github.com/spotify/helios/pull/618

**Pull request closed without merging**